src/main/scala/tech/sourced/gemini/ConnectedComponents.scala

```diff
@@ -0,0 +1,164 @@
+package tech.sourced.gemini
+
+import java.nio.ByteBuffer
+
+import com.datastax.driver.core.{Session, SimpleStatement}
+import org.slf4j.{Logger => Slf4jLogger}
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+case class FileHash(sha1: String, value: ByteBuffer)
+
+class DBConnectedComponents(
+                             log: Slf4jLogger,
+                             conn: Session,
+                             table: String,
+                             keyspace: String = Gemini.defautKeyspace)
+  extends ConnectedComponents(log) {
+  def getHashtables(): List[Byte] = {
+    val hashtablesCql = s"SELECT DISTINCT hashtable FROM $keyspace.$table"
+
+    conn
+      .execute(new SimpleStatement(hashtablesCql))
+      .asScala
+      .map(_.getByte("hashtable"))
+      .toList
+      .sorted
+  }
+
+  def getHashValues(hashtable: Byte): Iterable[FileHash] = {
+    val cql = s"SELECT sha1, value FROM $keyspace.$table WHERE hashtable=$hashtable"
+
+    conn.execute(new SimpleStatement(cql))
+      .asScala
+      .map(row => FileHash(row.getString("sha1"), row.getBytes("value")))
+  }
+}
+
+abstract class ConnectedComponents(log: Slf4jLogger) {
+  def getHashtables(): List[Byte]
+
+  def getHashValues(hashtable: Byte): Iterable[FileHash]
+
+  /**
+    * Group sha1-values from hashtables to buckets
+    *
+    * sequential sha1s with the same value in hashtable are be grouped to one bucket
+    * identical sha1s from different hashtables have the same element id
+    *
+    * it allows to operate with simple list of ints later instead of big ByteBuffers
+    *
+    * @return
+    */
+  def makeBuckets(): List[List[Int]] = {
+    val (buckets, elementIds) = getHashtables()
+      .foldLeft(List[List[Int]](), mutable.Map[String, Int]()) {
+        (result, hashtable) =>
```

Comment:
May be nice to be consistent with style of how function arguments are specified like i.e in https://github.com/smacker/gemini/blob/d96462356ac84906cb30fc1135f44d320701b604/src/main/scala/tech/sourced/gemini/ConnectedComponents.scala#L63